account_asset_date/models/account_asset.py

```diff
@@ -0,0 +1,165 @@
+from datetime import date, datetime
+from dateutil.relativedelta import relativedelta
+
+from odoo import api, models, fields
+from odoo.tools.translate import _
+from odoo.tools import DEFAULT_SERVER_DATE_FORMAT as DF
+from odoo.tools import float_is_zero
+
+
+class AccountAssetCategory(models.Model):
+    _inherit = 'account.asset.category'
+
+    different_dates = fields.Boolean('Set Start Depreciation Date',
+                                     default=False)
+
+
+class AccountAssetAsset(models.Model):
+    _inherit = 'account.asset.asset'
+    start_depreciation = fields.Date(
+        'Start Depreciation Date', help=_(
+            'Start Depreciation Date instead of Purchase Date'))
+    different_dates = fields.Boolean('Set Start Depreciation Date',
+                                     readonly=True,
+                                     states={'draft': [('readonly', False)]})
+
+    @api.model
+    def create(self, vals):
+        vals.update({
+            'start_depreciation': self._start_depreciation_date(
+                vals['date'], vals['prorata']),
+        })
+        asset = super(AccountAssetAsset, self).create(vals)
+        return asset
+
+    def _start_depreciation_date(self, deprec_date, prorata=False):
+        if prorata:
+            return (datetime.strptime(deprec_date, DF) +
+                    relativedelta(days=1)).strftime("%Y-%m-%d")
+        if deprec_date[8:] == '01':
+            return deprec_date
+        elif deprec_date[5:7] == '12':
+            return ('%d-01-01') % (int(deprec_date[:4]) + 1)
+        return ('%s-%02d-01') % (deprec_date[:4], int(
+            deprec_date[5:7]) + 1)
+
+    @api.multi
+    def compute_depreciation_board(self):
+        self.ensure_one()
+        if not self.different_dates:
+            return super(AccountAssetAsset, self).compute_depreciation_board()
+
+        posted_depreciation_line_ids = self.depreciation_line_ids.filtered(
+            lambda x: x.move_check).sorted(key=lambda l: l.depreciation_date)
+        unposted_depreciation_line_ids = self.depreciation_line_ids.filtered(
+            lambda x: not x.move_check)
+
+        # Remove old unposted depreciation lines. We cannot use unlink()
+        # with One2many field
+        commands = [(2, line_id.id, False) for line_id in
+                    unposted_depreciation_line_ids]
+
+        if self.value_residual != 0.0:
+            amount_to_depr = residual_amount = self.value_residual
+            if self.prorata:
+                # if we already have some previous validated entries, starting
+                # date is last entry + method perio
+                if (posted_depreciation_line_ids and
+                        posted_depreciation_line_ids[-1].depreciation_date):
```

Comment:
This condition is used twice. It could be better if we create a new variable(depreciation_date) to use it once.

```python
depreciation_date = posted_depreciation_line_ids and posted_depreciation_line_ids[-1].depreciation_date
```